torch_geometric/nn/models/rbcd_attack.py

```diff
@@ -0,0 +1,770 @@
+
+    def _append_statistics(self, mapping: Dict[str, Any]):
+        for key, value in mapping.items():
+            self.attack_statistics[key].append(value)
+
+    @torch.no_grad()
+    def _sample_final_edges(self, x: Tensor, labels: Tensor, budget: int,
+                            idx_attack: Optional[Tensor] = None,
+                            **kwargs) -> Tuple[Tensor, Tensor]:
+        best_metric = float('-Inf')
+        block_edge_weight = self.block_edge_weight
+        block_edge_weight[block_edge_weight <= self.coeffs['eps']] = 0
+
+        for i in range(self.coeffs['max_final_samples']):
+            if i == 0:
+                # In first iteration employ top k heuristic instead of sampling
+                sampled_edges = torch.zeros_like(block_edge_weight)
+                sampled_edges[torch.topk(block_edge_weight,
+                                         budget).indices] = 1
+            else:
+                sampled_edges = torch.bernoulli(block_edge_weight).float()
+
+            if sampled_edges.sum() > budget:
+                # Allowed budget is exceeded
+                continue
+            self.block_edge_weight = sampled_edges
+
+            edge_index, edge_weight = self._get_modified_adj(
+                self.edge_index, self.edge_weight, self.block_edge_index,
+                self.block_edge_weight)
+            prediction = self._forward(x, edge_index, edge_weight, **kwargs)
+            metric = self.metric(prediction, labels, idx_attack)
+
+            # Save best sample
+            if metric > best_metric:
+                best_metric = metric
+                best_edge_weight = self.block_edge_weight.clone().cpu()
+
+        # Recover best sample
+        self.block_edge_weight = best_edge_weight.to(self.device)
+        self.flipped_edges = self.block_edge_index[:,
+                                                   torch.where(best_edge_weight
+                                                               )[0]]
+
+        edge_index, edge_weight = self._get_modified_adj(
+            self.edge_index, self.edge_weight, self.block_edge_index,
+            self.block_edge_weight)
+        edge_mask = edge_weight == 1
+        self.edge_index = edge_index[:, edge_mask]
+        self.edge_weight = edge_weight[edge_mask]
+
+        return self.edge_index, self.flipped_edges
+
+    def _update_edge_weights(self, budget: int, epoch: int, gradient: Tensor):
+        # The learning rate is refined heuristically, s.t. (1) it is
+        # independent of the number of perturbations (assuming an undirected
+        # adjacency matrix) and (2) to decay learning rate during fine-tuning
+        # (i.e. fixed search space).
+        lr = (budget / self.n * self.lr /
+              np.sqrt(max(0, epoch - self.epochs_resampling) + 1))
+        self.block_edge_weight.data.add_(lr * gradient)
+
+    @staticmethod
+    def _num_possible_edges(n: int, is_undirected_graph: bool) -> int:
```

Comment:
I guess below here I could also move the methods into a "utils" file as DimeNet does, but I have no strong opinion.